components/constellation/constellation.rs

```diff
@@ -954,12 +955,37 @@ where
 
     /// The main event loop for the constellation.
     fn run(&mut self) {
-        while !self.shutting_down || !self.pipelines.is_empty() {
+        while !self.shutting_down {
             // Randomly close a pipeline if --random-pipeline-closure-probability is set
             // This is for testing the hardening of the constellation.
             self.maybe_close_random_pipeline();
             self.handle_request();
         }
+
+        // Try to cleanly exit all pipelines, until a timeout hits.
+        let pipeline_exit_timeout = after(Duration::from_millis(200));
+        loop {
+            if self.pipelines.is_empty() {
+                break;
+            }
+            select! {
+                recv(self.script_receiver) -> msg => {
+                    match msg.expect("Unexpected script channel panic in constellation") {
+                        Ok((source_pipeline_id, FromScriptMsg::PipelineExited)) => {
+                            self.handle_pipeline_exited(source_pipeline_id);
+                        },
+                        _ => continue,
+                    }
+                }
+                recv(pipeline_exit_timeout) -> _ => {
+                    if !self.pipelines.is_empty() {
+                        warn!("Failed to exit a number of pipelines {:?}", self.pipelines.len());
+                    }
```

Comment:
Is there any way to get more information about why it's blocked?

Reply:
I'm afraid we will have that warning and never pay attention to it.

Maybe make it an error, or dump a stack, not sure…

Reply:
I tried sampling them, but it's a bit messy because the bhm messaging stuff is not really setup for that, I think it could potentially introduce a new hang since we don't match bhm senders to pipeline yet, so perhaps best done as a follow-up. 

Also someone can use the dedicated sampling API to do this manually in some way, the current changes are more aimed at ensuring the window actually closes when someone wants to close it. 

Also this seems to mostly happen when JS is running and blocking the script-thread from handling the exit messages, so I think what we really need is a hook to kill the script-thread, see https://github.com/servo/servo/issues/27027